src/algebra/algebra/unitization.lean

```diff
@@ -0,0 +1,493 @@
+    by rw [smul_zero, zero_add, zero_smul, mul_zero, add_zero],
+  left_distrib := λ x₁ x₂ x₃, ext (mul_add x₁.1 x₂.1 x₃.1) $
+    show x₁.1 • (x₂.2 + x₃.2) + (x₂.1 + x₃.1) • x₁.2 + x₁.2 * (x₂.2 + x₃.2) =
+      x₁.1 • x₂.2 + x₂.1 • x₁.2 + x₁.2 * x₂.2 + (x₁.1 • x₃.2 + x₃.1 • x₁.2 + x₁.2 * x₃.2),
+    by { rw [smul_add, add_smul, mul_add], ac_refl },
+  right_distrib := λ x₁ x₂ x₃, ext (add_mul x₁.1 x₂.1 x₃.1) $
+    show (x₁.1 + x₂.1) • x₃.2 + x₃.1 • (x₁.2 + x₂.2) + (x₁.2 + x₂.2) * x₃.2 =
+      x₁.1 • x₃.2 + x₃.1 • x₁.2 + x₁.2 * x₃.2 + (x₂.1 • x₃.2 + x₃.1 • x₂.2 + x₂.2 * x₃.2),
+    by { rw [add_smul, smul_add, add_mul], ac_refl },
+  .. unitization.mul_one_class,
+  .. unitization.add_comm_monoid }
+
+instance [comm_monoid R] [non_unital_semiring A] [distrib_mul_action R A] [is_scalar_tower R A A]
+  [smul_comm_class R A A] : monoid (unitization R A) :=
+{ mul_assoc := λ x y z, ext (mul_assoc x.1 y.1 z.1) $
+    show (x.1 * y.1) • z.2 + z.1 • (x.1 • y.2 + y.1 • x.2 + x.2 * y.2) +
+      (x.1 • y.2 + y.1 • x.2 + x.2 * y.2) * z.2 =
+      x.1 • (y.1 • z.2 + z.1 • y.2 + y.2 * z.2) + (y.1 * z.1) • x.2 +
+      x.2 * (y.1 • z.2 + z.1 • y.2 + y.2 * z.2),
+    { simp only [smul_add, mul_add, add_mul, smul_smul, smul_mul_assoc, mul_smul_comm, mul_assoc],
+      nth_rewrite 1 mul_comm,
+      nth_rewrite 2 mul_comm,
+      abel },
+  ..unitization.mul_one_class }
+
+-- This should work for `non_unital_comm_semiring`s, but we don't seem to have those
+instance [comm_monoid R] [comm_semiring A] [distrib_mul_action R A] [is_scalar_tower R A A]
+  [smul_comm_class R A A] : comm_monoid (unitization R A) :=
+{ mul_comm := λ x₁ x₂, ext (mul_comm x₁.1 x₂.1) $
+    show x₁.1 • x₂.2 + x₂.1 • x₁.2 + x₁.2 * x₂.2 = x₂.1 • x₁.2 + x₁.1 • x₂.2 + x₂.2 * x₁.2,
+    by rw [add_comm (x₁.1 • x₂.2), mul_comm],
+  ..unitization.monoid }
+
+instance [comm_semiring R] [non_unital_semiring A] [module R A] [is_scalar_tower R A A]
+  [smul_comm_class R A A] : semiring (unitization R A) :=
+{ ..unitization.monoid,
+  ..unitization.non_assoc_semiring }
+
+-- This should work for `non_unital_comm_semiring`s, but we don't seem to have those
+instance [comm_semiring R] [comm_semiring A] [module R A] [is_scalar_tower R A A]
+  [smul_comm_class R A A] : comm_semiring (unitization R A) :=
+{ ..unitization.comm_monoid,
+  ..unitization.non_assoc_semiring }
+
+variables (R A)
+
+/-- The canonical inclusion of rings `R →+* unitization R A`. -/
+@[simps apply]
+def inl_ring_hom [semiring R] [non_unital_semiring A] [module R A] : R →+* unitization R A :=
+{ to_fun := inl,
+  map_one' := inl_one A,
+  map_mul' := inl_mul A,
+  map_zero' := inl_zero A,
+  map_add' := inl_add A }
+
+end mul
+
+section algebra
+variables (R A : Type*)
+variables [comm_semiring R] [non_unital_semiring A]
+variables [module R A] [is_scalar_tower R A A] [smul_comm_class R A A]
+
+instance algebra : algebra R (unitization R A) :=
```

Comment:
This should generalize to
```suggestion
instance algebra : algebra S (unitization R A) :=
```
Under appropriate assumptions on S

Reply:
done